src/proto/streams/streams.rs

```diff
@@ -622,25 +624,32 @@ where
 
 impl<B, P> Drop for StreamRef<B, P>
 where
+    B: Buf,
     P: Peer,
 {
     fn drop(&mut self) {
         let mut me = self.inner.lock().unwrap();
-
         let me = &mut *me;
 
-        let id = {
-            let mut stream = me.store.resolve(self.key);
-            stream.ref_dec();
+        let mut stream = me.store.resolve(self.key);
+        // decrement the stream's ref count by 1.
+        stream.ref_dec();
 
-            if !stream.is_released() {
-                return;
+        let actions = &mut me.actions;
+        // the reset must be sent inside a `transition` block.
+        // `transition_after` will release the stream if it is
+        // released.
+        me.counts.transition(stream, |_, stream|
+            // if this is the last reference to the stream,
+            // reset the stream.
+            if stream.ref_count == 0 {
+                actions.send.send_reset(
```

Comment:
I haven't looked in all the right places yet, but a reset shouldn't be sent if stream is already closed (because of EOS flags on frames sent and received). Does `send_reset` already take that into consideration? Or does it need to be done here?

Reply:
I believe so: https://github.com/carllerche/h2/blob/master/src/proto/streams/send.rs#L92-L101